### PR TITLE
Implement scoped pointers for gdal handled objects

### PR DESCRIPTION
**Pull request by nyalldawson**
Implements a c++11 unique_ptr with custom deleter for storing/automatic clean up of some gdal object types.

This fixes at least 2 leaks I noticed while implementing this, and I suspect it has the potential to save many more...

### CODE REVIEW

**Comment by m-kuhn** [comment]
Sounds good. What's left to do? I guess sqlite is another candidate.

**Comment by nyalldawson** [comment]
There's many more gdal objects which could benefit from these. Sqlite would be a great one too - I think we leak a lot with those.

File: src/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -1327,28 +1323,28 @@ bool QgsOgrProvider::addFeaturePrivate( QgsFeature &f, Flags flags )
 #ifdef OGRNullMarker
       OGR_F_SetFieldNull( feature, ogrAttId );
 #else
-      OGR_F_UnsetField( feature, ogrAttId );
+      OGR_F_UnsetField( feature.get(), ogrAttId );
```
Comment (rouault) [inline src/providers/ogr/qgsogrprovider.cpp]: missing .get() in #ifdef OGRNullMarker case

File: src/providers/ogr/qgsogrprovider.cpp [hunk 2]
```diff
@@ -1728,7 +1721,7 @@ bool QgsOgrProvider::changeAttributeValues( const QgsChangedAttributesMap &attr_
 #ifdef OGRNullMarker
         OGR_F_SetFieldNull( of, f );
 #else
-        OGR_F_UnsetField( of, f );
+        OGR_F_UnsetField( of.get(), f );
```
Comment (rouault) [inline src/providers/ogr/qgsogrprovider.cpp]: missing .get() in #ifdef OGRNullMarker case

**Review by rouault — CHANGES_REQUESTED** [review]
LGTM except the 2 cases I outlined

**Comment by nyalldawson** [comment]
>  What's left to do? I guess sqlite is another candidate.

Another good candidate: the gdal CPL objects. I suspect we leak a lot of those.

**Comment by nyalldawson** [comment]
@rouault how's d024910 for you?

**Comment by rouault** [comment]
> @rouault how's d024910 for you?

I've let a few comments.

**Comment by nyalldawson** [comment]
> I've let a few comments.

Thanks - much appreciated!

**Pull request merged**